### ISSUE TITLE
Ungraded Exercise 1 

### ISSUE THREAD

**Issue by Ramanenbh** [nus-cs2030/2223-s2#260]
I am stuck trying to figure out how to prevent cyclic dependency since Pager returns a Host and Transmitter return a Term. I have tried using intermediate classes between Host & Pager but this issue still seems to occur.

**Comment by briannoelkesuma** [comment]
Hi, I think u can try to limit the inheritance relationships. Doing so helped me to get rid of cyclic dependency.
E.g. TransmitterSnd not inheriting from Transmitter but extending from Host only.

**Comment by granolaluvr69** [comment]
It might be a little unintuitive, but one way to do it is to not have any transmitters inherit from other transmitters and no pagers inheriting from other pagers. However, they do all still need to inherit from host and term respectively, as they should always be instances of host and term respectively.

**Comment by 0liverkong** [comment]
additional stuff that may help:

snd() takes in a Host (in the form of Transmitter) but does not compile if you do something like snd(TransmitterSnd) (Level 5 requirement.) This means TransmitterSnd is part of the type which contains subtypes Transmitter, TransmitterSnd and TransmitterAck. Transmitter and TransmitterAck are a subtype which TransmitterSnd is not part of. This may hint you with the type of the argument snd takes in. It should be a new class defined by you.

**Comment by Ramanenbh** [comment]
Yup once i realised that TransSnd must not be a subtype of Transmitter but rather Host, the cyclic dependancy issue was eliminated. Im not exactly sure why but it works for me now.
Thanks for all the help guys :D
This is what i did for reference
![photo_2023-02-26_16-03-27](https://user-images.githubusercontent.com/124868539/221399278-39e12396-6e23-453e-bb83-57a43c38f807.jpg)

**Comment by kayeugenia** [comment]
> Yup once i realised that TransSnd must not be a subtype of Transmitter but rather Host, the cyclic dependancy issue was eliminated. Im not exactly sure why but it works for me now.

hi can I ask why must you create another Host Interface?

**Comment by granolaluvr69** [comment]
> > Yup once i realised that TransSnd must not be a subtype of Transmitter but rather Host, the cyclic dependancy issue was eliminated. Im not exactly sure why but it works for me now.
> 
> hi can I ask why must you create another Host Interface?

This is probably because we need to make it the case such that only complete handshakes can be accepted as arguments in the snd() method, so it makes us able to verify if the host passed into the snd() method has completed the handshake with another term or a new transmitter entirely, which has not established any connection yet. To restrict the types of hosts that can be passed into the snd() method, we can set the type of the argument taken in as the intermediate host interface.
